### PR TITLE
Add check for illegal Widnows names

### LINKED ISSUE #589: Add check for illegal Windows filenames
There are file and directory names which are legal in Git and most operating systems but illegal in Windows.  They should be avoided in case a developer happens to use Windows.

The following names are illegal as bare names as well as with extensions regardless of case for files or directories.  For instance `CoM6`, `cOm6.txt`, `com6.foo.jpg`, etc are all illegal.  Windows will refuse to create those files if you try to check them out.

From: https://docs.microsoft.com/en-us/windows/win32/fileio/naming-a-file

> Do not use the following reserved names for the name of a file:
> 
> CON, PRN, AUX, NUL, COM1, COM2, COM3, COM4, COM5, COM6, COM7, COM8, COM9, LPT1, LPT2, LPT3, LPT4, LPT5, LPT6, LPT7, LPT8, and LPT9. Also avoid these names followed immediately by an extension; for example, NUL.txt is not recommended. For more information, see [Namespaces](https://docs.microsoft.com/en-us/windows/win32/fileio/naming-a-file).

### PR DESCRIPTION
**Pull request by ericfrederich**
Fixes #589

### CODE REVIEW

**Comment by ericfrederich** [comment]
Took advice from #590 and created a new PR ... sorry for delay

File: .pre-commit-hooks.yaml [hunk 1]
```diff
@@ -40,6 +40,11 @@
     language: python
     types: [text, executable]
     stages: [commit, push, manual]
+-   id: check-illegal-windows-names
+    name: check illegal windows names
+    entry: Illegal windows filenames detected
+    language: fail
+    files: '(?i)(^|/)(CON|PRN|AUX|NUL|COM[1-9]|LPT[1-9])(\.|/|$)'
```
Comment (asottile) [inline .pre-commit-hooks.yaml:47]: can we add a quick little test which loads this regex from the configuration and runs it against a few expected values?

**Comment by ericfrederich** [comment]
Tests added, let me know if that's what you had in mind.

**Review by asottile — APPROVED** [review]

File: .pre-commit-hooks.yaml (hunk 1 above)
Comment (jleclanche) [reply .pre-commit-hooks.yaml:47]: @asottile @ericfrederich I recommend checking for the `:` character as well (see my comment in #589 regarding :Zone.Identifier files).

Comment (ericfrederich) [reply .pre-commit-hooks.yaml:47]: @jleclanche looks like this got merged without code checking for the `:` but no release has been made yet.

I just checked and it's definitely a problem on Windows.  I created a file `foo:bar` on linux, commited, pushed, then pulled down the repo from Windows and got greeted with `error: invalid path 'foo:bar'`

I'll add a test for it and a new PR

**Comment by thetillhoff** [comment]
Any plans on releasing a new version soon with this check included?

Just read about it on the Webpage, tried to add it, but it wasn't found. Then checked the dates, and the last release of pre-commit was before this PR was merged...

**Comment by asottile** [comment]
you're confused?

**Comment by thetillhoff** [comment]
Am I?
This PR adds the check-illegal-windows-names hook, and was merged on April 16th.
The last released version of this repo (https://github.com/pre-commit/pre-commit-hooks/releases/tag/v4.6.0) was released on April 6th.

Is there a plan to create a new release that includes the check-illegal-windows-names hook?
Or am I misunderstanding something?

**Comment by asottile** [comment]
have you read the pre-commit docs?

**Comment by thetillhoff** [comment]
Probably not the part you have in mind. You are welcome to point me to what you want me to read.

**Pull request merged**